packages/editor/src/components/post-author/index.js

```diff
@@ -1,115 +1,31 @@
-import { __ } from '@wordpress/i18n';
-import { ComboboxControl } from '@wordpress/components';
+import { withSelect } from '@wordpress/data';
+import { compose } from '@wordpress/compose';
 
 /**
  * Internal dependencies
  */
-import PostAuthorCheck from './check';
-
-function PostAuthor() {
-	const [ fieldValue, setFieldValue ] = useState();
-
-	const { authorId, isLoading, authors, postAuthor } = useSelect(
-		( select ) => {
-			const { __unstableGetAuthor, getAuthors, isResolving } = select(
-				'core'
-			);
-			const { getEditedPostAttribute } = select( 'core/editor' );
-			const author = __unstableGetAuthor(
-				getEditedPostAttribute( 'author' )
-			);
-			const query =
-				! fieldValue || '' === fieldValue ? {} : { search: fieldValue };
-			return {
-				authorId: getEditedPostAttribute( 'author' ),
-				postAuthor: author,
-				authors: getAuthors( query ),
-				isLoading: isResolving( 'core', 'getAuthors', [ query ] ),
-			};
-		},
-		[ fieldValue ]
-	);
-	const { editPost } = useDispatch( 'core/editor' );
-
-	const authorOptions = useMemo( () => {
-		const fetchedAuthors = ( authors ?? [] ).map( ( author ) => {
-			return {
-				value: author.id,
-				label: author.name,
-			};
-		} );
+import PostAuthorCombobox from './post-author-combobox';
+import PostAuthorSelect from './post-author-select';
 
-		// Ensure the current author is included in the dropdown list.
-		const foundAuthor = fetchedAuthors.findIndex(
-			( { value } ) => postAuthor?.id === value
-		);
+const minimumUsersForCombobox = 25;
 
-		if ( foundAuthor < 0 && postAuthor ) {
-			return [
-				{ value: postAuthor.id, label: postAuthor.name },
-				...fetchedAuthors,
-			];
-		}
-
-		return fetchedAuthors;
-	}, [ authors, postAuthor ] );
-
-	// Initializes the post author properly
-	// Also ensures external changes are reflected.
-	useEffect( () => {
-		if ( postAuthor ) {
-			setFieldValue( postAuthor.name );
-		}
-	}, [ postAuthor ] );
-
-	/**
-	 * Handle author selection.
-	 *
-	 * @param {number} postAuthorId The selected Author.
-	 */
-	const handleSelect = ( postAuthorId ) => {
-		if ( ! postAuthorId ) {
-			return;
-		}
-		editPost( { author: postAuthorId } );
-	};
-
-	/**
-	 * Handle user input.
-	 *
-	 * @param {string} inputValue The current value of the input field.
-	 */
-	const handleKeydown = ( inputValue ) => {
-		setFieldValue( inputValue );
-	};
-
-	if ( ! postAuthor ) {
-		return null;
+function PostAuthor( { authors } ) {
+	if ( authors?.length >= minimumUsersForCombobox ) {
+		return <PostAuthorCombobox />;
 	}
-
-	return (
-		<PostAuthorCheck>
-			<ComboboxControl
-				label={ __( 'Author' ) }
-				options={ authorOptions }
-				value={ authorId }
-				onFilterValueChange={ debounce( handleKeydown, 300 ) }
-				onChange={ handleSelect }
-				isLoading={ isLoading }
-				allowReset={ false }
-			/>
-		</PostAuthorCheck>
-	);
+	return <PostAuthorSelect />;
 }
 
-export default PostAuthor;
+export default compose( [
+	withSelect( ( select ) => {
```

Comment:
Should we be using useSelect instead?

Reply:
sure, I'll refactor this.

Reply:
Refactored PostAuthorSelect as a functional component, switching to `useSelect` and `useDispatch`.